crates/fluvio-spu-schema/src/server/api.rs

```diff
@@ -67,3 +67,38 @@ impl ApiMessage for SpuServerRequest {
         }
     }
 }
+
+impl SpuServerRequest {
+    pub fn correlation_id(&self) -> i32 {
+        match self {
+            Self::ApiVersionsRequest(request) => request.header.correlation_id(),
+            Self::ProduceRequest(request) => request.header.correlation_id(),
```

Comment:
So requires we have to implement this for every request.   Instead, do it this before: `decode_with_header`

Reply:
I don't think that we can instrument `decode_with_header` and get the result we want. Instrumentation follows a call stack, not a stream pipeline, which means the top-level instrumentation is at the top-level service handler, i.e. `handle_spu_request` in the SPU public service. I think it's unfortunate that we have to fan out like this since all the enum variants contain `RequestMessage<T>` with different types for `T`. I think this may indicate that there's a better way to structure this request format, but I didn't want to go that deep in trying to change things.

For now, I think this is a fixed-cost thing that we can improve later if we end up adding many more types of requests, but we have not added a new request type in a very long time so I think this is manageable, and I think the benefit to having `correlation_id` and `client_id` in the top-level spans is worth it.

Reply:
See my comment below.  If we have span start at the handler, we should not need this